[CIS-1136] Make it possible to encrypt/decrypt attachments

dmigach opened the pull request:
This change makes it possible to subclass our implementation of StreamCDNClient to implement their own
Also we make attachments mutable so it's possible to decrypt encrypted attachments
A customer gets names and avatarURLs for users from their own backend for security reasons, so they need to be able to set them.
Also this PR disables a flaky test

SagarSDagdu reviewed (dismissed):
LGTM!

evsaev commented on Sources/StreamChat/Config/ChatClientConfig.swift:
```diff
@@ -91,15 +91,20 @@ public struct ChatClientConfig {
     
-    /// Allows to inject a custom API client for uploading attachments, if not specified `StreamCDNClient` is used
-    public var customCDNClient: CDNClient?
+    public enum CDNClientConfig {
```
Would be great to have some docs here 👍

dmigach replied:
It seems like we might want to drop this change in the end if the customer will be able to achieve e2e without it

evsaev commented on Sources/StreamChat/Config/ChatClientConfig.swift:
```diff
@@ -91,15 +91,20 @@ public struct ChatClientConfig {
+        if case let .custom(client) = cdnClientConfig {
+            return type(of: client).maxAttachmentSize
         } else {
             return StreamCDNClient.maxAttachmentSize
```
Shouldn't we take a type from `cdnClientConfig` here? 🤔

evsaev replied:
would be great to also have a test for it 👍

evsaev commented on Sources/StreamChat/Models/Attachments/ChatMessageAudioAttachment.swift:
```diff
@@ -17,13 +17,13 @@ public struct AudioAttachmentPayload: AttachmentPayload {
+    public var file: AttachmentFile
     /// An extra data.
-    let extraData: [String: RawJSON]?
+    var extraData: [String: RawJSON]?
```
```suggestion
    public var extraData: [String: RawJSON]?
```

evsaev commented on Sources/StreamChat/Models/Attachments/ChatMessageFileAttachment.swift:
```diff
@@ -17,13 +17,13 @@ public struct FileAttachmentPayload: AttachmentPayload {
+    public var file: AttachmentFile
     /// An extra data.
-    let extraData: [String: RawJSON]?
+    var extraData: [String: RawJSON]?
```
```suggestion
    public var extraData: [String: RawJSON]?
```

evsaev commented on Sources/StreamChat/Models/Attachments/ChatMessageVideoAttachment.swift:
```diff
@@ -17,13 +17,13 @@ public struct VideoAttachmentPayload: AttachmentPayload {
+    public var file: AttachmentFile
     /// An extra data.
-    let extraData: [String: RawJSON]?
+    var extraData: [String: RawJSON]?
```
```suggestion
    public var extraData: [String: RawJSON]?
```

evsaev reviewed (commented):
Looks good, thanks! Let's address comments and we are good to 🚢

evsaev reviewed (commented):
Let's also have a changelog entry for this 🙏

dmigach commented:
A customer who requested these change is going to check if it helps them, if it does, we're going to finalise it. In the meantime I'll start addressing the issues 👌

nuno-vieira commented on Sources/StreamChat/Workers/ChannelUpdater_Tests.swift:
```diff
@@ -425,7 +425,8 @@ class ChannelUpdater_Tests: StressTestCase {
 
-    func test_hideChannel_successfulResponse_isPropagatedToCompletion() throws {
+    // TODO: Fix and enable
+    func _test_hideChannel_successfulResponse_isPropagatedToCompletion() throws {
```
Please remove this, it seems I was able to at least make it more stable on https://github.com/GetStream/stream-chat-swift/pull/1314

nuno-vieira commented:
> A customer who requested these change is going to check if it helps them, if it does, we're going to finalise it. In the meantime I'll start addressing the issues 👌

Any news about this one?

kojiba commented:
This decrease code readability cause when reader sees `var` he assumes its changing somewhere. 
Maybe there is options how to do this extendable but in more smart way, instead of just making everything public and var?

kojiba commented:
For example we could introduce a new delegate for ChannelController with will have a method 
`process(attachment: AttachmentInputType) -> AttachmentOutputType`
for input we can give a struct, and for output we could require just some proto to display data

nuno-vieira commented on Sources/StreamChat/Config/ChatClientConfig.swift:
```diff
@@ -171,3 +176,22 @@ public struct APIKey: Equatable {
 }
+
+public extension ChatClientConfig {
+    /// Allows to inject a custom API client for uploading attachments, if not specified `StreamCDNClient` is used
+    @available(*, deprecated, message: "Set `cdnClientConfig = .custom(client)` instead")
+    var customCDNClient: CDNClient? {
+        get {
+            guard case let .custom(client) = cdnClientConfig else { return nil }
+            
+            return client
+        }
+        set {
+            if let customClient = newValue {
+                cdnClientConfig = .custom(client: customClient)
+            } else {
+                cdnClientConfig = .default(type: StreamCDNClient.self)
+            }
+        }
+    }
```
No need to add deprecations, V4 will include breaking changes, so we can break the API for now 👍 Anyways, all deprecations should live in the `Deprecations.swift` file.

dmigach commented:
> > A customer who requested these change is going to check if it helps them, if it does, we're going to finalise it. In the meantime I'll start addressing the issues 👌
> 
> Any news about this one?

The customer was happy with the suggested solution, and they won't need to deal with CDN in the end. They went for implementing it so we might expect additional request in the nearest future
